Model/Api/DiscountCodeValidation.php

```diff
@@ -286,6 +286,11 @@ private function getCartTotals($quote)
     {
         $is_has_shipment = !empty($this->requestArray['cart']['shipments'][0]['reference']);
         $payload = $this->createPayloadForVirtualQuote($quote, $this->requestArray);
+        // if a quote is virtual and the payload has shipment it doesn't mean it is the payment only checkout
+        // virtual quote can have shipment if the quote was physical and then become to virtual in case with addons flow for example
+        if ($quote->getIsVirtual() && $is_has_shipment) {
```

Comment:
Checking `$is_has_shipment` is redundant. If it's true you set it to false, if it's already false then the result is the same.